src/typeChecker/__tests__/source2Typed.test.ts

```diff
@@ -0,0 +1,196 @@
+import { parseError } from '../..'
+import { mockContext } from '../../mocks/context'
+import { parse } from '../../parser/parser'
+import { Chapter, Variant } from '../../types'
+
+let context = mockContext(Chapter.SOURCE_2, Variant.TYPED)
+
+beforeEach(() => {
+  context = mockContext(Chapter.SOURCE_2, Variant.TYPED)
+})
+
+describe('null type', () => {
+  it('handles type mismatches correctly', () => {
+    const code = `const x1: null = null; // no error
+      const x2: null = '1'; // error
+      const x3: boolean = null; // error
+      const x4: undefined = null; // error
+      const x5: null = list(); // no error as null is empty list
+    `
+
+    parse(code, context)
+    expect(parseError(context.errors)).toMatchInlineSnapshot(`
+      "Line 2: Type 'string' is not assignable to type 'null'.
+      Line 3: Type 'null' is not assignable to type 'boolean'.
+      Line 4: Type 'null' is not assignable to type 'undefined'."
+    `)
+  })
+})
+
+describe('pair', () => {
+  it('handles type mismatches correctly', () => {
+    const code = `const x1: Pair<number, number> = pair(1, 2); // no error
+      const x2: Pair<number, number> = pair(1, '2'); // error
+      const x3: Pair<number, number> = 1; // error
+      const x4: Pair<number> = pair(1, 2); // error
+      const x5: Pair<number, number> = list(1, 2); // error
+    `
+
+    parse(code, context)
+    expect(parseError(context.errors)).toMatchInlineSnapshot(`
+      "Line 4: Generic type 'Pair' requires 2 type argument(s).
+      Line 2: Type 'Pair<number, string>' is not assignable to type 'Pair<number, number>'.
+      Line 3: Type 'number' is not assignable to type 'Pair<number, number>'.
+      Line 5: Type 'List<number>' is not assignable to type 'Pair<number, number>'."
+    `)
+  })
+
+  it('pair() must take in 2 arguments', () => {
+    const code = `pair(1, 2); // no error
+      pair(1); // error
+      pair(1, 2, 3); // error
+    `
+
+    parse(code, context)
+    expect(parseError(context.errors)).toMatchInlineSnapshot(`
+      "Line 2: Expected 2 arguments, but got 1.
+      Line 3: Expected 2 arguments, but got 3."
+    `)
+  })
+
+  it('lists are pairs', () => {
```

Comment:
Is this always true? What about the empty list?

Reply:
I don't think `null` is considered a pair, given that the definition of a list is "either `null` or a pair whose tail is a list". Will rename this for clarity and also add testcases for the case where `null` should not be typed as `Pair`.